### PR TITLE
Cluster + Dependency support

### PR DESCRIPTION
**Pull request by jrcrawfo**
## JIRA
https://ndustrialio.atlassian.net/browse/CONTXT-1667

## Why?
* Adding some commands to expose common support functions

## What changed?
- [x] Command to "get" clusters `contxt clusters get <slug>
- [x] Command to "register" a new cluster 
- [x] Command to get service dependencies 
- [x] Command to get service scopes
- [x] Command to add a service dependency
- [x] Command to add a scope to a dependency
- [x] Command to remove a scope from a dependency 
- [x] Resolve merge conflicts with new "click"-based CLI

### Services
`contxt services get` -> Get all services registered in Contxt
`contxt services get <service_id>` -> Get information about a specific service by its ID

### Service Dependencies
`contxt services dependencies get <service_id>` -> Get dependencies for a given service ID
`contxt services dependencies add --from-service-id <from_service> --to-service-id <to_service>` -> Create a new dependency between services
`contxt services dependencies add-scope --from-service-id <from_service> --to-service-id <to_service> --scopes read:example,write:example` -> Add scopes to an existing service dependency
`contxt services dependencies remove-scope  --from-service-id <from_service> --to-service-id <to_service> --scopes read:example,write:example` -> Remove scopes from an existing service dependency
`contxt services dependencies remove --from-service-id <from_service> --to-service-id <to_service>` -> Remove an existing dependency from a service

### Service Scopes
`contxt services scopes get <service_id>` -> Get all scopes for a service

### Clusters
`contxt clusters get --org "<org_name>" <cluster_slug> -> Get information about a specific cluster
`contxt clusters register --description "<Description>" --infrastructure-id <infra_id> --region "<aws_region>" --slug <slug>` -> Register a new cluster in Contxt to assign projects to

### CODE REVIEW

File: contxt/cli/commands/services.py [hunk 1]
```diff
@@ -1,23 +1,219 @@
...
+            print(f'Invalid scope "{scope}" -- does not exist in to_service')
+            continue
+        if scope not in existing_dep_scopes:
+            print(f"Scope {scope} is not currently granted in dependency -- doing nothing")
+        else:
+            to_remove.append(to_service_scopes[scope])
+
+    for scope in to_remove:
+        clients.contxt.remove_service_scope(to_dep, scope)
+        print(f"Removed {scope}")
```
Comment (branchvincent) [inline contxt/cli/commands/services.py:219]: thoughts on letting the api tell the user what is wrong rather than trying to exhaustively anticipate errors that the above loop is doing?

File: contxt/cli/commands/services.py [hunk 2]
```diff
@@ -1,23 +1,219 @@
...
+            continue
+        if scope not in existing_dep_scopes:
+            print(f"Adding scope {scope}")
+            to_add.append(to_service_scopes[scope])
+        else:
+            print(f"{scope} already exists for dependency")
+
+    for scope in to_add:
+        r = clients.contxt.add_service_scope(to_dep, scope)
+        print(r)
```
Comment (branchvincent) [inline contxt/cli/commands/services.py:188]: same comment as `remove_scope`

File: contxt/cli/commands/services.py [hunk 3]
```diff
@@ -1,23 +1,219 @@
...
+Scopes Commands
+"""
+
+
+@scopes.command("get")
+@click.argument("service_id")
+@fields_option(default=["label", "description"], obj=ServiceScope)
+@sort_option(default="label")
+@click.pass_obj
+def get_scopes(clients: Clients, service_id: str, fields: List[str], sort: str) -> None:
```
Comment (branchvincent) [inline contxt/cli/commands/services.py:81]: Can you add docstrings to these commands so we can get the description shown on `contxt services scopes get --help`, etc?

**Review by branchvincent — CHANGES_REQUESTED** [review]
couple requests + questions for my own knowledge.

also, can you share a list of complete CLI commands so anyone else reviewing can quickly run the new commands and see the output for themselves? for example:

```sh
contxt services get-scopes 1 # a real argument here
contxt services add-dep --service-id 1 --to-service-id 2 # ditto
```

**Comment by jrcrawfo** [comment]
> couple requests + questions for my own knowledge.
> 
> also, can you share a list of complete CLI commands so anyone else reviewing can quickly run the new commands and see the output for themselves? for example:
> 
> ```shell
> contxt services get-scopes 1 # a real argument here
> contxt services add-dep --service-id 1 --to-service-id 2 # ditto
> ```

Added these to the PR description. We will probably want to document these in some sort of documentation outside of the PR, but not sure where that would be quite yet.

File: contxt/services/contxt_deployments.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+from typing import List, Optional
+
+from ..auth import Auth
+from ..models.contxt import Cluster, ServiceGrant
+from .api import ApiEnvironment, ConfiguredApi
+
+
+class ContxtDeploymentService(ConfiguredApi):
```
[the remark this answers is not in the document]
Comment (jrcrawfo) [reply contxt/services/contxt_deployments.py:8]: Yeah, I would have liked to have done that, but the way the APIEnvironment class is structured right now, we don't really have a way to be compatible with some of the deploy vs. auth based routes we're using. For example, there are 3 route patterns being utilized by the Contxt API right now:
- contxt.api.ndustrial.io/deploy/v1/<other_routes> (Deployment/Tenant-related routes)
- contxt.api.ndustrial.io/access/v1/<other_routes> (Access-related routes)
- contxt.api.ndustrial.io/v1/<other_routes> ("Legacy"/Non-tenant-based routes)

The plan is to incrementally remove/move the "Legacy" routes into either of these two patterns since we want to make sure everything is divided by tenant.

Long story short, the way I decided to organize the services here was in service of that concept being created inside the Coordinator API (and I was kind of forced to because of how the ApiEnvironment class doesn't (currently) take into consideration where the "v1" needs to occur in the base_url. If I used the standard contxtService class, it would build all URLs as `/v1/<path>` when i actually need `deploy/v1/<path>`

File: contxt/services/contxt_deployments.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    """
+
+    def get_clusters(self, organization_id: str) -> List[Cluster]:
+        return [Cluster.from_api(rec) for rec in self.get(f"{organization_id}/clusters")]
+
+    def get_cluster(self, organization_id: str, cluster_slug: str) -> Cluster:
+        resp = self.get(f"{organization_id}/clusters/{cluster_slug}")
+        return Cluster.from_api(resp)
+
+    def register_cluster(
```
[the remark this answers is not in the document]
Comment (jrcrawfo) [reply contxt/services/contxt_deployments.py:33]: I normally would agree but I think in this case it's more accurate to call it register as the "Create" action of a cluster might get confusing since we're not actually creating an actual cluster, but rather registering that it does exist.

File: contxt/services/contxt_deployments.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            obj["secrets"] = {"BEARER_TOKEN": secret_bearer_token}
+
+        resp = self.post(f"{organization_id}/clusters", json=obj)
+        return Cluster.from_api(resp)
+
+    """
+    Dependencies
+    """
+
+    def remove_service_dependency(self, organization_id: str, service_grant: ServiceGrant) -> None:
```
[the remark this answers is not in the document]
Comment (jrcrawfo) [reply contxt/services/contxt_deployments.py:49]: Yeah, the nomenclature of the endpoint was derived by what Auth0 calls it, but "dependency" is what makes more sense for users, IMO.

File: contxt/models/contxt.py [hunk 7]
```diff
@@ -287,6 +288,7 @@ class Service(ApiObject):
     last_deployed_at: Optional[datetime]
     last_configured_at: Optional[datetime]
     service_type: str
+    stack_id: int
```
[the remark this answers is not in the document]
Comment (jrcrawfo) [reply contxt/models/contxt.py:291]: We'll need to do this whenever we merge in the new code and deploy to production. If we made this change now, this code wouldn't work. Where should we log the TODO here?

File: contxt/models/assets.py [hunk 8]
```diff
@@ -486,8 +486,8 @@ def metrics(self, metrics: Dict) -> None:
                     asset_id=self.asset.id,  # type: ignore
                     asset_metric_id=metric.id,  # type: ignore
                     effective_start_date=start_date,
-                    effective_end_date=self._effective_end_date(
-                        start_date, metric.time_interval
+                    effective_end_date=datetime.combine(
+                        self._effective_end_date(start_date, metric.time_interval), datetime.min.time()
```
[the remark this answers is not in the document]
Comment (jrcrawfo) [reply contxt/models/assets.py:490]: 🤔 good question. I was using this branch to pull some data (not realizing I was still on this branch) and fixed a broken endpoint related to pulling metric data.

File: contxt/cli/commands/clusters.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+@click.pass_obj
+def get(clients: Clients, org: str, cluster_slug: str):
+    organization = fetch_organization_from_name(clients, org)
+    cluster = clients.deployments.get_cluster(organization.id, cluster_slug)
+    print(Serializer.to_pretty_cli(cluster))
+
+
+@clusters.command()
+@click.option("--description", required=True)
+@click.option("--infrastructure_id", required=True)
```
[the remark this answers is not in the document]
Comment (jrcrawfo) [reply contxt/cli/commands/clusters.py:34]: Hmmm good question -- it looks like there is no "collections" endpoint unfortunately for getting infrastructures. There's only 1 GET endpoint, and that assume you already know the slug, which is kind of crappy.

File: contxt/cli/commands/services.py (hunk 1 above)
Comment (jrcrawfo) [reply contxt/cli/commands/services.py:219]: I'm usually in the camp of handling things like this with "prettier" warnings specific to the command rather than some ugly exceptions coming from the API with no context of whether it's expected or not. If we didn't do this, we would see some 404 exceptions and we aren't really giving the user much context around whether it's something they need to be concerned with or not.

**Comment by jrcrawfo** [comment]
Was absorbed by a separate effort
